core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -171,12 +173,15 @@ private[spark] class CoarseMesosSchedulerBackend(
     synchronized {
       val filters = Filters.newBuilder().setRefuseSeconds(5).build()
 
-      for (offer <- offers) {
+      // filter out all the offers that do not meet constraints (if specified)
+      val qualifyingOffers = filterOffersByConstraints(offers, slaveOfferConstraints)
```

Comment:
Thanks for that tidbit. I was indeed leaving them hanging around. Another thing which i noticed at: https://github.com/apache/spark/pull/5563/files#diff-d425d35aa23c47a62fbb538554f2f2cfR213

Instead of declining offers the code replies with a `launchTasks` with no taskInfo. Does anyone know which one is preferred / correct / better.